uphold/vault-assist-tool
# Add support for xrpl tokens

### New features
- Ability to withdraw all XPRL tokens to another account 
- Recover XRP reserve from token trust lines
- Ability to withdraw available XRP when other tokens have not been removed
- Fix error handling for invalid XRP destination account

LABS-788
LABS-789
LABS-790
LABS-791
LABS-792
LABS-795
LABS-794
LABS-793
LABS-796
LABS-797
LABS-798
LABS-799

### Testing instructions
To start with testnet:
```
NET=development yarn dev
```

To start with mainnet:
```
yarn start
```

Also be sure to check the new e2e tests:
```
yarn e2e
```

##### Basic steps

1. Start the web server the correct environment (above)
2. Open browser to `http://localhost:3000/`
3. Make sure you have an XRPL multisig created, and then follow the steps below for each scenario.
4. Be sure that the mutlisig is funded by following the instructions in each scenario, including adding or removing funds/tokens to meet the requirements.
5. Make sure that you have two additional accounts to use as a destination. One account needs to be funded with XRP, and another account needs to be funded with the token you are testing.

##### Withdraw XRPL tokens with the notice screen
1. Requirement: Start with a multisig that is funded with both XRP and XRPL tokens.
2. Select `XRP` from the Asset type
3. Enter the vault address and continue
4. You will be given a notice to select withdrawing available `XRP` or a specific token.
5. Choose a specific token to withdraw
6. Error handling: Proceed through the steps, entering the destination address of an account that only has XRP, and you should see an error that the address is invalid
6. Enter the destination address of your secondary account that already has the token trust line
7. Confirm that the amounts and fees look correct.
8. Submit transaction, confirm "view transaction" opens the correct explorer link
9. Confirm no other issues with fees or amounts

##### Withdraw XRPL tokens by choosing it up front
1. Requirement: Start with a multisig that is funded with both XRP and XRPL tokens.
2. Select a specific XRPL token from the Asset type
3. Enter the vault address and continue
4. Error handling: Proceed through the steps, entering the destination address of an account that only has XRP, and you should see an error that the address is invalid
5. Enter the destination address of your secondary account that already has the token trust line
6. Confirm that the amounts and fees look correct.
7. Submit transaction, confirm "view transaction" opens the correct explorer link
8. Confirm no other issues with fees or amounts

##### Withdraw XRP tokens and without closing account (trust lines present)
1. Requirement: Start with a multisig that is funded with both XRP and XRPL tokens.
4. Select `XRP` from the Asset type
3. Enter the vault address and continue
5. Enter the destination address of your secondary account that has been funded
6. Confirm that the amounts and fees look correct.
7. Submit transaction, the transaction success should show that vault remains active
8. Confirm "view transaction" opens the correct explorer link
9. Confirm no other issues with fees or amounts


##### Withdraw XRPL tokens when they hold no balance (close trust line)
1. Requirement: Start with a multisig that funded with XRP, and add a token trust line but do not deposit any tokens (zero balance)
2. Select a specific XRPL token from the Asset type
3. Enter the vault address and continue
5. Enter the destination address of your secondary account that already has the token trust line
6. Confirm that the amounts and fees look correct
7. Submit transaction, the transaction success should show that the trust line is closed and the reserve is credited
8. Confirm "view transaction" opens the correct explorer link
10. Confirm no other issues with fees or amounts

##### Withdraw XRP tokens and close account (error testing)
1. Requirement: Start with a multisig that funded with XRP, but does not have any token trust lines or hold any token balances
2. Select a specific XRPL token from the Asset type (even though the vault does not have any)
3. Error handling: Enter the vault address and continue, an error should occur because the vault does not support that token
4. Select `XRP` from the Asset type
5. Error handling: Proceed through the steps, entering the destination address of an account that has not been created or funded, and you should see an error that the address is invalid
5. Enter the destination address of your funded secondary account
6. Confirm that the amounts and fees look correct
7. Submit transaction, confirm "view transaction" opens the correct explorer link
9. Confirm no other issues with fees or amounts

**NOTE**: For good measure, we should make sure BTC functions in vault-assist-tool still work.
